wpilibj/src/main/java/edu/wpi/first/wpilibj/apriltag/AprilTagFieldLayout.java

```diff
@@ -0,0 +1,140 @@
+ *
+ * <p>The JSON format contains a top-level "tags" field, which is a list of all AprilTags contained
+ * within a layout. Each AprilTag serializes to a JSON object containing an ID and a Pose3d.
+ *
+ * <p>Pose3ds are assumed to be measured from the bottom-left corner of the field, when the blue
+ * alliance is at the left. Pose3ds will automatically be returned as passed in when calling {@link
+ * AprilTagFieldLayout#getTagPose(int)}. Setting an alliance color via {@link
+ * AprilTagFieldLayout#setAlliance(DriverStation.Alliance)} will mirror the poses returned from
+ * {@link AprilTagFieldLayout#getTagPose(int)} to be correct relative to the other alliance.
+ */
+@JsonIgnoreProperties(ignoreUnknown = true)
+@JsonAutoDetect(getterVisibility = JsonAutoDetect.Visibility.NONE)
+public class AprilTagFieldLayout {
+  @JsonProperty(value = "tags")
+  private final List<AprilTag> m_apriltags = new ArrayList<>();
+
+  private boolean m_mirror;
+
+  /**
+   * Construct a new AprilTagFieldLayout with values imported from a JSON file.
+   *
+   * @param path Path of the JSON file to import from.
+   * @throws IOException If reading from the file fails.
+   */
+  public AprilTagFieldLayout(String path) throws IOException {
+    this(Path.of(path));
+  }
+
+  /**
+   * Construct a new AprilTagFieldLayout with values imported from a JSON file.
+   *
+   * @param path Path of the JSON file to import from.
+   * @throws IOException If reading from the file fails.
+   */
+  public AprilTagFieldLayout(Path path) throws IOException {
+    this(new ObjectMapper().readValue(path.toFile(), AprilTagFieldLayout.class).m_apriltags);
+  }
+
+  /**
+   * Construct a new AprilTagFieldLayout from a list of {@link AprilTag} objects.
+   *
+   * @param apriltags List of {@link AprilTag}.
+   */
+  @JsonCreator
+  public AprilTagFieldLayout(
+      @JsonProperty(required = true, value = "tags") List<AprilTag> apriltags) {
+    // To ensure the underlying semantics don't change with what kind of list is passed in
+    m_apriltags.addAll(apriltags);
+  }
+
+  /**
+   * Set the alliance that your team is on.
+   *
+   * <p>This changes the {@link #getTagPose(int)} method to return the correct pose for your
+   * alliance.
+   *
+   * @param alliance The alliance to mirror poses for.
+   */
+  public void setAlliance(DriverStation.Alliance alliance) {
+    m_mirror = alliance == DriverStation.Alliance.Red;
+  }
+
+  /**
+   * Gets an AprilTag pose by its ID.
+   *
+   * @param ID The ID of the tag.
+   * @return The pose corresponding to the ID passed in.
+   */
+  @SuppressWarnings("ParameterName")
+  public Pose3d getTagPose(int ID) {
+    Pose3d pose = m_apriltags.stream().filter(it -> ID == it.m_ID).findFirst().get().m_pose;
+    if (m_mirror) {
+      pose =
+          pose.relativeTo(
+              new Pose3d(
+                  new Translation3d(Units.feetToMeters(54.0), Units.feetToMeters(27.0), 0.0),
```

Comment:
Maybe inject this in? Its usually 27x54 but their have been minor differences according to the manuals and pathweaver files

2014: `24 ft. 8 in. x 54 ft`
2017: ` 27 ft. by 54 ft. 4 in`
2020: `26 ft. 11¼ in. (~821 cm) by 52 ft. 5¼ in`

Reply:
Yep, good point.

Reply:
Added, I believe the commits added resolve this issue.  May have also accidentally fixed another few issues along the way so that's good.